### PR TITLE
Add testing for Pytorch instance group kind MODEL

### PR DESCRIPTION
**Pull request by krishung5**
Pytorch backend: https://github.com/triton-inference-server/pytorch_backend/pull/107

### CODE REVIEW

**Comment by krishung5** [comment]
Unfortunately `torch._C._cuda_sleep` is not supported in TorchScript:
```
Python builtin <built-in function _cuda_sleep> is currently not supported in Torchscript:
  File "/opt/tritonserver/qa/L0_libtorch_instance_group_kind_model/gen_models.py", line 37
    def forward(self, x):
        torch._C._cuda_sleep(10)
        ~~~~~~~~~~~~~~~~~~~~ <--- HERE
        return torch.sum(x, dim=1
```
I couldn't find a way to make a torch model sleep and occupy on GPU as well.

**Review by Tabrizian — COMMENTED** [review]
The testing looks good, could you also add the same model in L0_io as well?

**Review by tanmayv25 — APPROVED** [review]

File: qa/L0_libtorch_instance_group_kind_model/gen_models.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+
+class DiffModule(nn.Module):
+
+    def __init__(self, device):
+        super(DiffModule, self).__init__()
+        self.device = device
+
+    def forward(self, INPUT0, INPUT1):
+        INPUT0 = INPUT0.to(self.device)
```
Comment (tanmayv25) [inline qa/L0_libtorch_instance_group_kind_model/gen_models.py:53]: Does this ensure that execution will also occur on the same device? Is it possible that execution kernels to be invoked on the default device access the tensors via p2p? 
Can you verify this with nvidia-smi?

Comment (krishung5) [reply qa/L0_libtorch_instance_group_kind_model/gen_models.py:53]: I ran PA on the model that uses device 0 and device 2, and the output of nvidia-smi showed that only the two devices were used:
```
krish@nvdl-a112-asus01:/$ nvidia-smi
Mon Jun 12 13:56:05 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 515.65.01    Driver Version: 515.65.01    CUDA Version: 11.7     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A40          On   | 00000000:01:00.0 Off |                    0 |
|  0%   46C    P0    81W / 300W |    730MiB / 46068MiB |      3%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A40          On   | 00000000:41:00.0 Off |                    0 |
|  0%   45C    P0    79W / 300W |    728MiB / 46068MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  NVIDIA A40          On   | 00000000:81:00.0 Off |                    0 |
|  0%   44C    P0    78W / 300W |    730MiB / 46068MiB |      2%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  NVIDIA A40          On   | 00000000:C1:00.0 Off |                    0 |
|  0%   45C    P0    80W / 300W |    730MiB / 46068MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A     11479      C   tritonserver                      728MiB |
|    1   N/A  N/A     11479      C   tritonserver                      726MiB |
|    2   N/A  N/A     11479      C   tritonserver                      728MiB |
|    3   N/A  N/A     11479      C   tritonserver                      728MiB |
+-----------------------------------------------------------------------------+
```
>Is it possible that execution kernels to be invoked on the default device access the tensors via p2p?

I think the default device should be used for accessing the tensor. We can see that the default device, which is GPU 0, has more utilization. Also, for a model that uses CPU and device 3, we can see that device 0 is used as well:
```
krish@nvdl-a112-asus01:/$ nvidia-smi
Mon Jun 12 13:58:45 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 515.65.01    Driver Version: 515.65.01    CUDA Version: 11.7     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A40          On   | 00000000:01:00.0 Off |                    0 |
|  0%   47C    P0    81W / 300W |    730MiB / 46068MiB |      1%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A40          On   | 00000000:41:00.0 Off |                    0 |
|  0%   46C    P0    80W / 300W |    728MiB / 46068MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  NVIDIA A40          On   | 00000000:81:00.0 Off |                    0 |
|  0%   45C    P0    78W / 300W |    730MiB / 46068MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  NVIDIA A40          On   | 00000000:C1:00.0 Off |                    0 |
|  0%   46C    P0    81W / 300W |    730MiB / 46068MiB |      2%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A     11479      C   tritonserver                      728MiB |
|    1   N/A  N/A     11479      C   tritonserver                      726MiB |
|    2   N/A  N/A     11479      C   tritonserver                      728MiB |
|    3   N/A  N/A     11479      C   tritonserver                      728MiB |
+-----------------------------------------------------------------------------+
```
Let me try to the nsight traces. Had some issue with nsight hanging when generating the report.

Comment (krishung5) [reply qa/L0_libtorch_instance_group_kind_model/gen_models.py:53]: Resolved the nsight hanging issue and finally got the traces. I used the `libtorch_multi_gpu` testing model that used device 0 and device 2. From the traces we can see that the kernels were launched on both of device 0 and device 2.
image

Comment (tanmayv25) [reply qa/L0_libtorch_instance_group_kind_model/gen_models.py:53]: Thanks for confirming Kris!

**Pull request merged**